### PR TITLE
test: access keys scenarios in meta transactions

### PR DESCRIPTION
**Pull request by jakmeier**
Final set of tests to complement the integration tests listed in #8467.

### CODE REVIEW

File: integration-tests/src/tests/client/features/delegate_action.rs [hunk 1]
```diff
@@ -287,6 +291,172 @@ fn meta_tx_fn_call() {
...
+    // Check allowance was not updated
+    let key =
+        node.user().get_access_key(&sender, &public_key).expect("failed looking up fn access key");
+    let AccessKeyPermissionView::FunctionCall { allowance, ..} = key.permission else {
+        panic!("should be function access key")
+    };
+    assert_eq!(
+        allowance.unwrap(),
+        initial_allowance,
+        "allowance should not change, we used the relayer's fund not the sender's"
```
Comment (nagisa) [inline integration-tests/src/tests/client/features/delegate_action.rs:358]: Should this also check relayer’s and sender’s balance to validate that relayer’s funds been used and that sender’s overall balance was also left untouched (regardless of what the kay’s allowance says?)

File: integration-tests/src/tests/client/features/delegate_action.rs [hunk 2]
```diff
@@ -287,6 +291,172 @@ fn meta_tx_fn_call() {
...
+    );
+}
+
+/// Call a function in a meta tx where the user only has access through a
+/// function call access that has zero allowance left.
```
Comment (nagisa) [inline integration-tests/src/tests/client/features/delegate_action.rs:363]: The comment does not match the implementation (`initial_allowance = 1`.)

File: integration-tests/src/tests/client/features/delegate_action.rs [hunk 3]
```diff
@@ -287,6 +291,172 @@ fn meta_tx_fn_call() {
...
+    let node = RuntimeNode::new_from_genesis(&relayer, genesis);
+
+    let actions = vec![Action::FunctionCall(FunctionCallAction {
+        method_name,
+        args: vec![],
+        gas: 30_000_000_000_000,
+        deposit: 0,
+    })];
+
+    let tx_result = node.user().meta_tx(sender, receiver, relayer, actions).unwrap();
```
Comment (nagisa) [inline integration-tests/src/tests/client/features/delegate_action.rs:442]: There's an unfortunate amount of duplication between the three tests added here? Would it be possible to consolidate them somehow?

Comment (jakmeier) [reply integration-tests/src/tests/client/features/delegate_action.rs:405]: I tried my best to reduce duplication, but the consolidated function need an awful lot of parameters. See my latest commit.

In terms of LOC, I ended up with more code than before. But I think it's still an improvement, as the logical duplication is reduced. Not sure if the code is cleaner to read though.

File: integration-tests/src/tests/client/features/delegate_action.rs (hunk 2 above)
Comment (jakmeier) [reply integration-tests/src/tests/client/features/delegate_action.rs]: should be fixed now

File: test-utils/testlib/src/runtime_utils.rs [hunk 4]
```diff
@@ -50,3 +51,19 @@ pub fn add_contract(genesis: &mut Genesis, account_id: &AccountId, code: Vec<u8>
     }
     records.push(StateRecord::Contract { account_id: account_id.clone(), code });
 }
+
+/// Add an account with a specified access key to the genesis state records.
+pub fn add_account_with_access_key(
+    genesis: &mut Genesis,
+    sender: &AccountId,
```
Comment (mm-near) [inline test-utils/testlib/src/runtime_utils.rs:58]: nit: not sure why you call it 'sender' and not 'account_id'

File: test-utils/testlib/src/runtime_utils.rs [hunk 5]
```diff
@@ -50,3 +51,19 @@ pub fn add_contract(genesis: &mut Genesis, account_id: &AccountId, code: Vec<u8>
     }
     records.push(StateRecord::Contract { account_id: account_id.clone(), code });
 }
+
+/// Add an account with a specified access key to the genesis state records.
```
Comment (mm-near) [inline test-utils/testlib/src/runtime_utils.rs:55]: Access key & balance

File: integration-tests/src/tests/client/features/delegate_action.rs [hunk 6]
```diff
@@ -572,3 +715,36 @@ fn assert_ft_balance(
...
+    add_account_with_access_key(&mut genesis, &sender, NEAR_BASE, public_key, access_key);
+    RuntimeNode::new_from_genesis(user, genesis)
+}
+
+fn fn_access_key(
+    initial_allowance: u128,
+    receiver_id: String,
+    method_names: Vec<String>,
+) -> AccessKey {
+    let access_key = AccessKey {
```
Comment (mm-near) [inline integration-tests/src/tests/client/features/delegate_action.rs:741]: nit: local variable not needed.

**Review by mm-near — APPROVED** [review]

**Pull request merged**